riscv/src/continuations/bootloader.rs

```diff
@@ -22,6 +22,26 @@ pub const MEMORY_HASH_START_INDEX: usize = 2 * REGISTER_NAMES.len();
 pub const NUM_PAGES_INDEX: usize = MEMORY_HASH_START_INDEX + 8;
 pub const PAGE_INPUTS_OFFSET: usize = NUM_PAGES_INDEX + 1;
 
+/// Computes an upper bound of how long the shutdown routine will run, for a given number of pages.
+pub fn num_shutdown_routine_upper_bound(num_pages: usize) -> usize {
+    // Regardless of the number of pages, we have to:
+    // - Jump to the start of the routine
+    // - Assert all register values are correct (except the PC)
+    // - Start the page loop
+    // - Jump to shutdown sink
+    let constant_overhead = 6 + REGISTER_NAMES.len() - 1;
+
+    // For each page, we have to:
+    // - Start the page loop (14 instructions)
+    // - Load all words of the page
+    // - Invoke the hash function once every 4 words
+    // - Assert the page hash is as claimed (4 instructions)
+    // - Increment the page index and jump back to the loop start (2 instructions)
+    let cost_per_page = 14 + WORDS_PER_PAGE + WORDS_PER_PAGE / 4 + 4 + 2;
```

Comment:
these things feel kind of fragile, but fine for now

Reply:
Absolutely. With the current approach, we never run the shutdown routine in the RISC-V executor, so we can't really measure how long it will take.

Running it the executor would be complicated, because:
- The RISC-V executor would also have to support the "jump to shutdown routine" mechanism, which is quite specific...
- Even if it had that, we'd still have to set the value to pass to the executor. Without any estimate, all we could do would be to just try until we find one that works (completes the shutdown routine in time but otherwise doesn't waste too much computation).

On the upside, if the estimate is slightly too large, that's fine, it will only waste a bit of time. If the estimate is too low, the full constraints implemented in #915 would fail because the PC in the last row is constrained to be in equal to `shutdown_sink`, which means the shutdown routine must have completed.

Reply:
I guess one thing that could work is to run the executor but with the initial PC (passed via bootloader inputs) set to the start of the shutdown routine.

Assuming that the runtime of the shutdown routine does not depend on the memory content (it doesn't, but that's an assumption), this would let us measure the runtime exactly.

However, we would also have to change:
- The prover-provided page hashes at the end of the execution though, so that the checks in the shutdown routine actually pass.
- If we change that, we also have to change the Merkle proofs for the updates, so that the checks in the bootloader pass.
  - Alternatively, if the executor has a way to initialize the PC directly, we can bypass the bootloader.

So possible, but kinda complicated.